subprojects/plugin-development/src/main/java/org/gradle/plugin/devel/plugins/ExternalPluginValidationPlugin.java

```diff
@@ -0,0 +1,143 @@
+        project.getPlugins().configureEach(p -> {
+            Class<?> pluginClass = p.getClass();
+            if (isExternal(pluginClass)) {
+                registry.findPluginForClass(pluginClass).ifPresent(pluginId -> {
+                    File pluginJar = findPluginJar(pluginClass);
+                    if (pluginJar != null) {
+                        List<File> pluginJars = jarsByPlugin.get(pluginId);
+                        if (pluginJars == null) {
+                            pluginJars = Lists.newArrayList();
+                            TaskProvider<ValidatePlugins> validationTask = configureValidationTask(project, pluginJars, pluginId);
+                            lifecycleTask.configure(t -> t.dependsOn(validationTask));
+                            jarsByPlugin.put(pluginId, pluginJars);
+                        }
+                        pluginJars.add(pluginJar);
+                    }
+                });
+            }
+        });
+    }
+
+    private static PluginRegistry findPluginRegistry(Project project) {
+        return ((ProjectInternal) project).getServices().get(PluginRegistry.class);
+    }
+
+    private static TaskProvider<Task> createLifecycleTask(Project project) {
+        return project.getTasks().register(VALIDATE_EXTERNAL_PLUGINS_TASK_NAME);
+    }
+
+    private TaskProvider<ValidatePlugins> configureValidationTask(Project project,
+                                                                         List<File> pluginJars,
+                                                                         PluginId pluginId) {
+        String idWithoutDots = pluginId.getId().replace('.', '_');
+        return project.getTasks().register(VALIDATE_EXTERNAL_PLUGIN_TASK_PREFIX + idWithoutDots, ValidatePlugins.class, task -> {
+            task.setGroup(PLUGIN_DEVELOPMENT_GROUP);
+            task.setDescription(VALIDATE_PLUGIN_TASK_DESCRIPTION);
+            task.getOutputFile().set(project.getLayout().getBuildDirectory().file("reports/plugins/validation-report-for-" + idWithoutDots + ".txt"));
+
+            ScriptHandlerInternal scriptHandler = (ScriptHandlerInternal) project.getBuildscript();
+            List<File> scriptClassPath = scriptHandler.getScriptClassPath().getAsFiles();
+            task.getClasses().setFrom(pluginClassesOf(pluginJars));
+            task.getClasspath().setFrom(scriptClassPath);
+        });
+    }
+
+    private List<FileTree> pluginClassesOf(List<File> pluginJars) {
+        return pluginJars.stream()
+            .map(fileOperations::zipTree)
+            .collect(Collectors.toList());
+    }
+
+    @Nullable
+    private static File findPluginJar(Class<?> pluginClass) {
+        return toSourceFile(pluginClass.getProtectionDomain().getCodeSource().getLocation());
+    }
+
+    private static boolean isExternal(Class<?> pluginClass) {
+        return !pluginClass.getName().startsWith("org.gradle");
+    }
+
+    @Nullable
+    private static File toSourceFile(URL url) {
```

Comment:
The name "source file" is a misleading for me. This is a JAR file that the class was loaded from, right?